saleor/payment/gateways/adyen/utils.py

```diff
@@ -124,19 +85,79 @@ def request_data_for_payment(
             "currency": payment_information.currency,
         },
         "reference": payment_information.payment_id,
-        "paymentMethod": payment_data.get("paymentMethod"),
+        "paymentMethod": payment_method,
         "returnUrl": return_url,
         "merchantAccount": merchant_account,
         **extra_request_params,
     }
 
+    method = payment_method.get("type", [])
     if "klarna" in method:
-        # TODO
-        append_klarna_data(payment_information, request_data)
-
+        request_data = append_klarna_data(payment_information, request_data)
     return request_data
 
 
+def append_klarna_data(payment_information: "PaymentData", payment_data: dict):
+    _type, payment_pk = graphene.Node.from_global_id(payment_information.payment_id)
+    checkout = Checkout.objects.filter(payments__id=payment_pk).first()
+
+    if not checkout:
+        raise PaymentError("Unable to calculate products for klarna")
```

Comment:
Do we handle this error in API somehow? Wondering when it can happen and what the user should do in this case.

Reply:
CheckoutComplete handles it.